Add support for MAX31875

jemfgeronimo opened the pull request:
## PR Description

- This adds driver support and device tree bindings for MAX31875 Temperature Sensor.
- Datasheet: [MAX31875](https://www.analog.com/media/en/technical-documentation/data-sheets/MAX31875.pdf)
- Tested on RPI4 with [MAX31875EVKIT](https://www.analog.com/en/resources/evaluation-hardware-and-software/evaluation-boards-kits/max31875evkit.html#eb-overview)

## PR Type
- [ ] Bug fix (a change that fixes an issue)
- [x] New feature (a change that adds new functionality)
- [ ] Breaking change (a change that affects other repos or cause CIs to fail)

## PR Checklist
- [x] I have conducted a self-review of my own code changes
- [ ] I have tested the changes on the relevant hardware
- [ ] I have updated the documentation outside this repo accordingly (if there is the case)

nunojsa reviewed (commented):
First round of review from me...

jemfgeronimo commented on Documentation/devicetree/bindings/hwmon/adi,max31875.yaml:
```diff
@@ -0,0 +1,75 @@
+  adi,comp-int:
+    description:
+      If present interrupt mode is used. If not present comparator mode is used
+      (default).
```
I based this on an existing dt property: [adi,comp-int](https://github.com/analogdevicesinc/linux/blob/893fbaa78dcdd19b39d68cbf4e2034657ec7144a/Documentation/devicetree/bindings/hwmon/adi%2Cmax31827.yaml#L35)

What this controls is the behavior of a STATUS BIT of the device: whether it will be de-asserted via an I2C read or when going below the hysteresis value. This is documented [here](https://github.com/analogdevicesinc/linux/blob/9f5d6a3ae108917b42ba83d6b7ca7991e86cca05/Documentation/hwmon/max31875.rst?plain=1#L29-L40).

Maybe, I can just modify the description to better reflect this behavior.

nunojsa replied:
Hmm, I see... interrupt mode is misleading :)

jemfgeronimo replied:
I will change the description into

      If present, the Overtemperature Status bit operates in interrupt mode. If
      not present, it operates in comparator mode (default).

jemfgeronimo commented on drivers/hwmon/max31875.c:
```diff
@@ -0,0 +1,559 @@
+		.name = "max31875",
+		.of_match_table = max31875_of_match,
+	},
+	.probe_new = max31875_probe,
```
Noted, thanks!

jemfgeronimo commented on drivers/hwmon/max31875.c:
```diff
@@ -0,0 +1,559 @@
+	&dev_attr_temp1_resolution.attr,
+	NULL
+};
+ATTRIBUTE_GROUPS(max31875);
```
I just copied from a similar device: [max31827](https://github.com/analogdevicesinc/linux/blob/f313cf09b8e6b2aa389e4946c894a4861a78500e/drivers/hwmon/max31827.c#L486).

I honestly don't know the need of making this a runtime configurable other than the general flexibility. I can set this resolution in dt. What do you think?

nunojsa replied:
Well, if there's precedence already upstream for this then fine by me... Up to the maintainer to decide :). But now that I briefly looked at max31827 I saw some similarities. Did you made sure that it's not feasible to support this device in that driver?

jemfgeronimo replied:
In terms of features, I would say that max31875 is like a subset of max31827. I have considered adding in max31827,c the support for max31875 but it will include mismatch of register addresses, bit fields, data format, resolution (e.g., 8bits in max31875 already includes sign bit, 8bits in max31827 not yet includes sign bit), etc. that will involve adding many conditional statements that will make the code more complex and complicated.

In short, in my opinion, adding support for max31875 in max31827.c is not straightforward enough.

jemfgeronimo replied:
I now created a separated PR for this: https://github.com/analogdevicesinc/linux/pull/2635

jemfgeronimo commented on drivers/hwmon/max31875.c:
```diff
@@ -0,0 +1,559 @@
+
+	*val *= sign ? -1 : 1;
+
+	*val = ((*val << (extend ? 7 : 8)) / MILLI) * (sign ? -1 : 1);
```
For some reason, the math is only correct when I need to make `val` nonnegative first before I do the math then apply `sign` afterwards.

I will try to test this again in HW, maybe I can do this differently.

jemfgeronimo commented on drivers/hwmon/max31875.c:
```diff
@@ -0,0 +1,559 @@
+{
+	int ret;
+	unsigned int uval;
+
```
Maybe I can just put a scoped_guard on cases where I need the lock:

		case hwmon_temp_input:
			scoped_guard(mutex, &data->lock)
				ret = max31875_read_temp(data, true,
							 MAX31875_TEMP_REG,
							 val);

			return ret;
		case hwmon_temp_max:
			return max31875_read_temp(data, false,
						  MAX31875_T_OS_REG, val);
		case hwmon_temp_max_hyst:
			return max31875_read_temp(data, false,
						  MAX31875_T_HYST_REG, val);

jemfgeronimo replied:
I don't need the lock on every call of `max31875_read_temp`.

Also, I need advice on whether to use `guard(mutex)(&lock)` or `scoped_guard(mutex, &lock)`.

jemfgeronimo replied:
I think I've figure it out. Some errors were fixed while trying it in HW.

jemfgeronimo commented on drivers/hwmon/max31875.c:
```diff
@@ -0,0 +1,559 @@
+					 FIELD_PREP(mask, res));
+
+	if (!ret && idx)
+		*idx = res;
```
I would change that to:

	if (ret)
		return ret;

	if (idx)
		*idx = res;

	return 0;

Since, `idx` can be `NULL`.

jemfgeronimo commented on drivers/hwmon/max31875.c:
```diff
@@ -0,0 +1,559 @@
+	return ret;
+}
+
+static inline void max31875_convert_raw_to_temp(u8 resolution, unsigned int raw,
```
Which one is this referring to?

jemfgeronimo commented on drivers/hwmon/max31875.c:
```diff
@@ -0,0 +1,559 @@
+	unsigned int uval;
+
+	if (is_temp_input) {
+		if (!data->enable) {
```
This was also copied from a similar driver: [max31827.c](https://github.com/analogdevicesinc/linux/blob/f313cf09b8e6b2aa389e4946c894a4861a78500e/drivers/hwmon/max31827.c#L223-L239).
From what I've understood, the `hwmon_temp_enable` attr is used there to toggle the device between shutdown mode (disabled) and autonomous conversion mode (enable). In shutdown mode, the device can still capture a temperature measurement via one-shot function. This one-shot function is what you see in this portion of the code.

Now, what I can do to comply with [https://www.kernel.org/doc/Documentation/hwmon/sysfs-interface](https://www.kernel.org/doc/Documentation/hwmon/sysfs-interface) is to remove the one-shot in the driver and just return `-ENODATA` when `hwmon_temp_input` is read in shutdown mode (`hwmon_temp_enable = 0`).

nunojsa replied:
No, if there's precedence, leave it like this... But makes me again wonder if we could not support this device in the other driver

jemfgeronimo commented:
V2
- added `adi,extended-format`
- modified `adi,comp-int` description
- changed `adi,timeout-enable` to `adi,timeout-disable`
- changed `adi,fault-q` from `u8` to `u32`
- renamed `resolution` of `max31875_data` to `bit_reso`
- added `bool extend_fmt` to `max31875_data`
- renamed `shutdown_write` to `max31875_shutdown_write`
- removed ugly cast from a function call of `max31875_update_bits`
- fixed indentions/alignments
- refactored `max31875_convert_raw_to_temp()` and `max31875_convert_temp_to_raw()`
- used switch cases to handle the `type` and `attr` in `max31875_is_visible()`, `max31875_read()` and `max31875_write()`
- removed `i2c_set_clientdata()` call

nunojsa commented on drivers/hwmon/max31875.c:
```diff
@@ -0,0 +1,635 @@
+
+static const struct of_device_id max31875_of_match[] = {
+	{ .compatible = "adi,max31875" },
+	{ },
```
no need to the comma on the last delimiter entry...

nunojsa commented on drivers/hwmon/max31875.c:
```diff
@@ -0,0 +1,635 @@
+
+static const struct i2c_device_id max31875_id[] = {
+	{ "max31875", },
+	{ },
```
ditto

nunojsa commented on drivers/hwmon/max31875.c:
```diff
@@ -0,0 +1,635 @@
+	if (!data)
+		return -ENOMEM;
+
+	mutex_init(&data->lock);
```
can use devm_mutex_init() when upstreaming

jemfgeronimo replied:
Noted, thanks!

nunojsa commented on drivers/hwmon/max31875.c:
```diff
@@ -0,0 +1,635 @@
+	int ret;
+
+	if (!i2c_check_functionality(client->adapter, I2C_FUNC_SMBUS_WORD_DATA))
+		return -EOPNOTSUPP;
```
not sure if the above is needed given that devm_regmap_init_i2c() will do it's own checks

nunojsa commented on drivers/hwmon/max31875.c:
```diff
@@ -0,0 +1,635 @@
+			ret = max31875_update_bits(data, false,
+					max31875_conversions,
+					ARRAY_SIZE(max31875_conversions),
+					(const u16)val, MAX31875_CONV_RATE_MASK,
```
Why the cast? No need for it...

nunojsa commented on drivers/hwmon/max31875.c:
```diff
@@ -0,0 +1,635 @@
+					(const u16)val, MAX31875_CONV_RATE_MASK,
+					NULL);
+			if (!ret)
+				data->update_interval = (u16)val;
```
Ditto. If `update_interval` is u16 then the cast is already implicit. It seems you're also missing some locking.  max31875_update_bits() should be serialized with setting `data->update_interval`. Otherwise you can do the device write but someone sneaks in and reads the variable before being updated (while the HW is already configured for the new value).

nunojsa commented on drivers/hwmon/max31875.c:
```diff
@@ -0,0 +1,635 @@
+			if (val >> 1)
+				return -EOPNOTSUPP;
+
+			scoped_guard(mutex, &data->lock) {
```
Not sure if it's available already but this is being improved upstream. Basically you'll be able to do:

```
if (ret)
    return ret;

data->enable = val;
return 0;
```

All inside the guard block and the compiler won't complain about it...

jemfgeronimo replied:
Noted. Thanks!

nunojsa commented on drivers/hwmon/max31875.c:
```diff
@@ -0,0 +1,635 @@
+
+			ret = max31875_write_temp(data, MAX31875_T_OS_REG, val);
+			if (!ret)
+				data->temp_os = val;
```
possible you have two calls to max31875_write_temp() so you need a lock (plus writing in shared variables). Also the typical pattern in linux is to handle errors first. Hence:

```
if (ret)
   return ret;

data->temp_os = val;
return 0;
```

nunojsa commented on drivers/hwmon/max31875.c:
```diff
@@ -0,0 +1,635 @@
+			ret = max31875_write_temp(data, MAX31875_T_HYST_REG,
+						  val);
+			if (!ret)
+				data->temp_hyst = val;
```
ditto about pattern and locking

nunojsa commented on drivers/hwmon/max31875.c:
```diff
@@ -0,0 +1,635 @@
+						  MAX31875_T_OS_REG, val);
+		case hwmon_temp_max_hyst:
+			return max31875_read_temp(data, false,
+						  MAX31875_T_HYST_REG, val);
```
Why `hwmon_temp_input` has the lock and the other's don't? I guess you know what you're doing but it this is a bit confusing. And this is not a fastpath at all so I would keep things simple by moving the lock inside the max31875_read_temp() .

nunojsa commented on drivers/hwmon/max31875.c:
```diff
@@ -0,0 +1,635 @@
+					return ret;
+
+				*val = !FIELD_GET(MAX31875_SHUT_DOWN_MASK,
+						  uval);
```
No need for locking. You're just doing a regmap_read() which has a lock already.

nunojsa commented on drivers/hwmon/max31875.c:
```diff
@@ -0,0 +1,635 @@
+		}
+	default:
+		return 0;
+	}
```
nit: typically this function comes before the read/write ones...

jemfgeronimo replied:
I'm confused. It is already before `max31875_read()` and `max31875_write()` or am I misunderstanding it?

nunojsa commented on drivers/hwmon/max31875.c:
```diff
@@ -0,0 +1,635 @@
+	unsigned int uval;
+
+	if (data->extend_fmt)
+		return -EOPNOTSUPP;
```
You can do better than this :). If this extra attribute depends on a DT parameter, just  don't add it if `data->extend_fmt`.

nunojsa commented on drivers/hwmon/max31875.c:
```diff
@@ -0,0 +1,635 @@
+	 */
+	if (data->extend_fmt)
+		return sysfs_emit(buf, "%u\n",
+				  max31875_resolutions[MAX31875_RES_12_BIT]);
```
Alright, disregard my previous comment. I see we can still read it. You could still conditionally do something with the attr permissions but yeah, not worth it.

nunojsa commented on drivers/hwmon/max31875.c:
```diff
@@ -0,0 +1,635 @@
+	*val = (*val << (extend_fmt ? 7 : 8)) / (long)MILLI;
+}
+
+static int max31875_write_temp(struct max31875_data *data, u8 reg, long val)
```
If I'm not missing something `data` could be made `const`

nunojsa commented on drivers/hwmon/max31875.c:
```diff
@@ -0,0 +1,635 @@
+
+static int max31875_write_temp(struct max31875_data *data, u8 reg, long val)
+{
+	max31875_convert_temp_to_raw(data->extend_fmt, &val);
```
This is only called here. I would drop the helper function and just do the conversion here...

nunojsa commented on drivers/hwmon/max31875.c:
```diff
@@ -0,0 +1,635 @@
+	/**
+	 * In the extended format, the MSB is increased from 64C to 128C.
+	 */
+	*val = (*val << (extend_fmt ? 7 : 8)) / (long)MILLI;
```
Why the cast? Maybe also consider `DIV_ROUND_CLOSEST()`

nunojsa commented on drivers/hwmon/max31875.c:
```diff
@@ -0,0 +1,635 @@
+	if (ret)
+		return ret;
+
+	max31875_convert_raw_to_temp(data->extend_fmt, uval, val);
```
ditto... I would just do this inline in here

nunojsa commented on drivers/hwmon/max31875.c:
```diff
@@ -0,0 +1,635 @@
+	 */
+	s32 temp = (sign_extend32(raw, 15) * (long)MILLI) >> (extend_fmt ? 7 : 8);
+
+	*val = temp;
```
I think this could be a one liner... Just assign directly to *val

nunojsa commented on drivers/hwmon/max31875.c:
```diff
@@ -0,0 +1,635 @@
+	/**
+	 * In the extended format, the MSB is increased from 64C to 128C.
+	 */
+	s32 temp = (sign_extend32(raw, 15) * (long)MILLI) >> (extend_fmt ? 7 : 8);
```
do we need the cast to long? I think what you need to reason is that if the multiplication can actually lead to 64bits? If so, you need a cast to s64...

nunojsa commented on drivers/hwmon/max31875.c:
```diff
@@ -0,0 +1,635 @@
+			ret = regmap_write(data->regmap, reg, val);
+		else
+			ret = regmap_update_bits(data->regmap, reg, mask, val);
+		return ret;
```
nit: Slightly better:

```
if (!mask)
    return regmap_write(data->regmap, reg, val);

return regmap_update_bits(data->regmap, reg, mask, val);
```

nunojsa commented on drivers/hwmon/max31875.c:
```diff
@@ -0,0 +1,635 @@
+	 *
+	 * Mutex is used to ensure, that some other process doesn't change the
+	 * configuration register.
+	 */
```
No need for locking comment. It's a bit redundant

nunojsa reviewed (commented):
Some more inputs from me
